pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -582,43 +612,51 @@ private boolean buildSegmentInternal(boolean forCommit) {
     } catch (IOException e) {
       segmentLogger.error("Exception during move/tar segment", e);
       FileUtils.deleteQuietly(tempSegmentFolder);
-      return false;
+      return null;
     }
     FileUtils.deleteQuietly(tempSegmentFolder);
     long endTimeMillis = System.currentTimeMillis();
 
     _serverMetrics.setValueOfTableGauge(_metricKeyName, ServerGauge.LAST_REALTIME_SEGMENT_CREATION_DURATION_SECONDS,
         TimeUnit.MILLISECONDS.toSeconds(endTimeMillis - startTimeMillis));
 
-    return true;
+    if (forCommit) {
+      return destDir.getAbsolutePath() + TarGzCompressionUtils.TAR_GZ_FILE_EXTENTION;
+    }
+    return destDir.getAbsolutePath();
   }
 
-  protected boolean commitSegment() {
+  protected boolean commitSegment(final String segTarFileName) {
     // Send segmentCommit() to the controller
     // if that succeeds, swap in-memory segment with the one built.
-    File destSeg = makeSegmentDirPath();
-    final String segTarFileName = destSeg.getAbsolutePath() + TarGzCompressionUtils.TAR_GZ_FILE_EXTENTION;
-    try {
-      SegmentCompletionProtocol.Response response = _protocolHandler.segmentCommit(_currentOffset, _segmentNameStr,
-          new File(segTarFileName));
-      if (!response.getStatus().equals(SegmentCompletionProtocol.ControllerResponseStatus.COMMIT_SUCCESS)) {
-        segmentLogger.warn("Received controller response {}", response);
-        return false;
-      }
-      _realtimeTableDataManager.replaceLLSegment(_segmentNameStr, _indexLoadingConfig);
-    } catch (FileNotFoundException e) {
-      segmentLogger.error("Tar file {} not found", segTarFileName, e);
+    File segTarFile = new File(segTarFileName);
+    if (!segTarFile.exists()) {
+      throw new RuntimeException("Segment file does not exist:" + segTarFileName);
+    }
+    SegmentCompletionProtocol.Response response =  postSegmentCommitMsg(segTarFile);
+    if (!response.getStatus().equals(SegmentCompletionProtocol.ControllerResponseStatus.COMMIT_SUCCESS)) {
+      segmentLogger.warn("Received controller response {}", response);
+      // Save the segment file if commit fails
```

Comment:
Just use _segmentFileAndOffset throughout, set it on build, reset on successful commit or consumption